src/Crypto/Hash/PoseidonNative.hs

```diff
@@ -0,0 +1,3228 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+
+module Crypto.Hash.PoseidonNative (poseidon) where
+
+import Data.List(foldl')
+import qualified Data.Primitive.Array as Array
+import qualified Data.Primitive.SmallArray as SmallArray
+
+modulus :: Integer
+modulus = 21888242871839275222246405745257275088548364400416034343698204186575808495617
+
+mulmod :: Integer -> Integer -> Integer
+mulmod a b = (a * b) `mod` modulus
+
+summod :: Integer -> Integer -> Integer
+summod a b = (a + b) `mod` modulus
+
+sig :: Integer -> Integer
+sig inVal =
+  let in2 = mulmod inVal inVal
+      in4 = mulmod in2 in2
+  in mulmod in4 inVal
```

Comment:
My group theory is rusty, but can you do `mod` just once after all the multiplications?

Reply:
I tried some basic tests and that seems to hold true:
```
module Main where

modulus :: Integer
modulus = 53

mulmod :: Integer -> Integer -> Integer
mulmod a b = (a * b) `mod` modulus

sig :: Integer -> Integer
sig inVal =
  let in2 = mulmod inVal inVal
      in4 = mulmod in2 in2
  in mulmod in4 inVal

mulmod2:: Integer -> Integer -> Integer
mulmod2 a b = (a * b)

sig2 :: Integer -> Integer
sig2 inVal =
  let in2 = mulmod2 inVal inVal
      in4 = mulmod2 in2 in2
  in mulmod2 in4 inVal `mod` modulus

main :: IO ()
main = do
    print $ sig 123
    print $ sig 456
    print $ sig 789
    print $ sig2 123
    print $ sig2 456
    print $ sig2 789
```

Reply:
Alright, I grabbed a pen and paper today! So the question effectively is whether `ab mod n = (a mod n) × (b mod n) mod n`. Let's use brute force and assume `a = k₁n + r₁`, `b = k₂n + r₂` (with the usual conditions of `r₁, r₂ < n`). Then:
```
ab mod n = (...) × n + r₁r₂ mod n = r₁r₂ mod n
```
while for the rhs
```
(a mod n) × (b mod n) mod n = r₁r₂ mod n
```
so that clearly holds, and `mod` can happen just once. That is assuming nothing funny happens wrt negative signs here and haskell `mod` behaves like math `mod` (which IIRC it does).

Reply:
Thanks Georg, worth a try.

Just tried it, it's not any faster according to my benchmark to do 
```hs
(inVal ^ 5) `mod` modulus
```
than to do what we're doing now. I think that's what you were suggesting.

Reply:
Yep, that was it.

Well, perhaps it's not that much of a hotstop then! Up to you what expression to keep, although I'd probably vote for `inVal ^ 5` — this way it's more obvious what's happening.